src/realm/sync/noinst/client_reset.cpp

```diff
@@ -353,111 +461,115 @@ void client_reset::transfer_group(const Transaction& group_src, Transaction& gro
+                bool nullable = col_key.is_nullable();
                 bool has_search_index = table_src->has_search_index(col_key);
                 logger.trace("Create column, table = %1, column name = %2, "
                              " type = %3, nullable = %4, has_search_index = %5",
-                             table_name, col_name, type, nullable, has_search_index);
+                             table_name, col_name, col_key.get_type(), nullable, has_search_index);
                 ColKey col_key_dst;
-                if (Table::is_link_type(ColumnType(type))) {
+                if (Table::is_link_type(col_key.get_type())) {
                     ConstTableRef target_src = table_src->get_link_target(col_key);
                     TableRef target_dst = group_dst.get_table(target_src->get_name());
-                    if (type == type_LinkList) {
+                    if (col_key.is_list()) {
                         col_key_dst = table_dst->add_column_list(*target_dst, col_name);
                     }
+                    else if (col_key.is_set()) {
+                        col_key_dst = table_dst->add_column_set(*target_dst, col_name);
+                    }
+                    else if (col_key.is_dictionary()) {
+                        DataType key_type = table_src->get_dictionary_key_type(col_key);
+                        col_key_dst = table_dst->add_column_dictionary(*target_dst, col_name, key_type);
+                    }
                     else {
+                        REALM_ASSERT(!col_key.is_collection());
                         col_key_dst = table_dst->add_column(*target_dst, col_name);
                     }
                 }
-                else if (col_key.get_attrs().test(col_attr_List)) {
-                    col_key_dst = table_dst->add_column_list(type, col_name, nullable);
+                else if (col_key.is_list()) {
+                    col_key_dst = table_dst->add_column_list(col_type, col_name, nullable);
+                }
+                else if (col_key.is_set()) {
+                    col_key_dst = table_dst->add_column_set(col_type, col_name, nullable);
+                }
+                else if (col_key.is_dictionary()) {
+                    DataType key_type = table_src->get_dictionary_key_type(col_key);
+                    col_key_dst = table_dst->add_column_dictionary(col_type, col_name, nullable, key_type);
                 }
                 else {
-                    col_key_dst = table_dst->add_column(type, col_name, nullable);
+                    REALM_ASSERT(!col_key.is_collection());
+                    col_key_dst = table_dst->add_column(col_type, col_name, nullable);
                 }
 
                 if (has_search_index)
                     table_dst->add_search_index(col_key_dst);
             }
+            else {
+                // column preexists in dest, make sure the types match
+                if (col_key.get_type() != col_key_dst.get_type() ||
```

Comment:
I think this is not consistent with the checks found higher up where the destination columns were removed if type etc. did not match. Perhaps the above checks should be removed so that we just fail the reset if columns with same name do not match.